kotlin/src/main/kotlin/net/jqwik/kotlin/api/CombinatorDsl.kt

```diff
@@ -0,0 +1,97 @@
+package net.jqwik.kotlin.api
+
+import net.jqwik.api.Arbitrary
+import org.apiguardian.api.API
+import java.util.Optional
+import kotlin.properties.ReadOnlyProperty
+import kotlin.reflect.KProperty
+
+
+/**
+ * Combine arbitraries using the combinator DSL:
+ *
+ * ```kotlin
+ * combine {
+ *     val first by Arbitraries.strings()
+ *     val second by Arbitraries.strings()
+ *
+ *     createAs {
+ *         "first: $first, second: $second"
+ *     }
+ * }
+ * ```
+ *
+ * @return new Arbitrary instance
+ */
+@API(status = API.Status.EXPERIMENTAL, since = "1.8.0")
+fun <R> combine(
+    combinator: CombinatorScope.() -> Combined<R>
+): Arbitrary<R> {
+    val combined = CombinatorScope().combinator()
+
+    @Suppress("UNCHECKED_CAST")
+    return combine(combined.arbitraries as List<Arbitrary<Any?>>) { values ->
+        try {
+            for ((property, value) in combined.properties zip values) {
+                property.bind(value)
+            }
+
+            CombinatorCreateScope().(combined.combinator)()
```

Comment:
I wonder what are the overheads of the approach. Basicaly you have to create and discard several thread locals for every "combine" call.

I wonder if you can use a single thread-local to pass `list` of the values and access items by the corresponding index. Each property can remember its index, and their get could be `getListOfBoundValues.get(index)`.

WDYT?

Reply:
I guess you add `CombinatorCreateScope()` for "improvements in the future", so it might make sense to cache a single object for now to avoid dummy allocations on every call

Reply:
I originally had a global `ThreadLocal`, but I could potentially see issues with some (admittedly very weird) nested calls of `combine`:

```kotlin
fun outer(): Arbitrary<String> {
    return combine {
        val value by Arbitraries.strings()

        createAs {
            value + nested { nestedValue -> nestedValue + value } .sample()
                                                       // ^ the get call for this 
                                                       //   value would now refer
                                                       //   to the list of values
                                                       //   from nested
        }
    }
}

fun nested(mapper: (String) -> String): Arbitrary<String> {
    return combine {
        val value by Arbitraries.strings()

        createAs { mapper(value) }
    }
}
```

Of course this example is absolutely ridiculous, but I think a _global_ `ThreadLocal` could cause some very weird issues.

What should be possible is creating a single `ThreadLocal` inside `combine`, passing it to all properties that are created. Because then, the `ThreadLocal` wouldn't ever be overwritten. Or is that what you are referring to?

Reply:
What I mean is "every `combine { ... }` uses its own `TheradLocal` for the list of values.
Of course, nested `combine { ... combine { ... }  }` calls would get two independent thread locals.